closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -474,12 +562,257 @@ JNIEXPORT jlong JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_loadCrypto
+            if (tempVersion > 0) {
+                return result;
+            }
+        }
+    }
+
+    /* The attempt to load from property and OpenSSL bundled with JDK failed.
+     * Try loading the libraries in the order set out above, and retain the latest library.
+     */
+    for (i = 0; i < numOfLibs; i++) {
+        if (traceEnabled) {
+            fprintf(stdout, "Attempting to load libname from OS : %s\n", libNames[i]);
+        }
+        result = load_crypto_library_libname(traceEnabled, libNames[i]);
+
+        if (NULL == result) {
+            continue;
+        }
+
+        /* Identify and load the latest version from the available libraries.
+         * This logic depends upon the order in which libnames are defined.
+         * It only loads the libraries which can possibly be the latest versions.
+         */
+        log_crypto_library_path(traceEnabled, result, "\tLibrary to be potentially used was loaded from");
+        tempVersion = get_crypto_library_version(traceEnabled, result, "\tLoaded OpenSSL version");
+
+        if (tempVersion <= 0) {
+            continue;
+        }
+
+        if (i >= num_of_generic) {
+            /* If library checked is not a generic one, there is no need to check further. */
+            if (previousVersion > tempVersion) {
+                unload_crypto_library(result);
+                return prevResult;
+            } else {
+                if (previousVersion != 0) {
+                    unload_crypto_library(prevResult);
+                }
+                return result;
+            }
+        } else {
+             /* If library checked is a generic one, we need to keep checking for higher versions. */
+            if (tempVersion > previousVersion) {
+                previousVersion = tempVersion;
+                prevResult = result;
+            }
+        }
+    }
+
+    /* If we reach here, it means that none of the non-generic libraries
+     * where found. However, a generic one might have been found in the
+     * process and, if so, it will be in the prevResult variable.
+     */
+    return prevResult;
+}
+
+/*
+ * Class:     jdk_crypto_jniprovider_NativeCrypto
+ * Method:    loadCrypto
+ * Signature: (ZLjava/lang/String;Ljava/lang/String;)J
+ */
+JNIEXPORT jlong JNICALL
+Java_jdk_crypto_jniprovider_NativeCrypto_loadCrypto
+  (JNIEnv * env, jobject jobj, jboolean traceEnabled, jstring jlibname, jstring jhomepath)
+{
+    const char *clibname = "";
+    const char *chomepath = "";
+    jlong ossl_ver = 0;
+
+    if (NULL != jlibname) {
+        clibname = (*env)->GetStringUTFChars(env, jlibname, NULL) ;
+        if (clibname[0] == '\0') {
+            if (traceEnabled) {
+                fprintf(stderr, "The jdk.native.openssl.lib property is not set.\n");
+                fflush(stderr);
+            }
+        } else {
+            crypto_library = load_crypto_library_libname(traceEnabled, (const char *)clibname);
```

Comment:
The cast is redundant here; please remove it.

Reply:
Removed.